File: Go/Day02.go
package main

import AH "./adventhelper"

import (
	"strings"
	"strconv"
)

type Instruction struct {
	Dir   string
	Value int
}

type Pos struct {
	V int
	H int
	Aim int
}

func parseLine(s string) Instruction {
	parts := strings.Split(s, " ")
	value, _ := strconv.Atoi(parts[1])

	return Instruction{Dir:parts[0], Value:value}
}

func (p *Pos) apply(i Instruction) () {
	switch i.Dir {
		case "forward": p.H += i.Value
		case "up":      p.V -= i.Value
		case "down":    p.V += i.Value

		default: // do nothing
	}	

	return
}

func (p *Pos) applyWithAim(i Instruction) () {
	switch i.Dir {
		case "forward":
			p.H += i.Value
			p.V += p.Aim * i.Value
		case "up": p.Aim -= i.Value
		case "down": p.Aim += i.Value

		default: // do nothing
	}	

	return
}

func main() {
	ss, _ := AH.ReadStrFile("../input/input02.txt")

	p1 := Pos{H:0, V:0}
	p2 := Pos{H:0, V:0}
	for _, s := range ss {
		i := parseLine(s)
		p1.apply(i)
		p2.applyWithAim(i)
	}

	AH.PrintSoln(2, p1.H * p1.V, p2.H * p2.V)

	return
}

File: Haskell/Day02.hs
import AdventHelper

fst' (x, _, _) = x
snd' (_, x, _) = x

parseRule :: String -> (String, Int)
parseRule s = (takeWhile(/= ' ') s, i)
  where i = read $ drop 1 $ dropWhile(/= ' ') s :: Int

apply :: (Int, Int) -> (String, Int) -> (Int, Int)
apply (h, v) (d, val)
  | d == "forward" = (h + val, v)
  | d == "up"      = (h, v - val)
  | d == "down"    = (h, v + val)

applyWithAim :: (Int, Int, Int) -> (String, Int) -> (Int, Int, Int)
applyWithAim (h, v, a) (d, val)
  | d == "forward" = (h + val, v + (a * val), a)
  | d == "up"      = (h, v, a - val)
  | d == "down"    = (h, v, a + val)

main = do
  putStrLn "Day 2"
  f <- readFile "../input/input02.txt"
  let is = map parseRule $ lines f

  let p1 = foldl apply (0,0) is 
  let p2 = foldl applyWithAim (0,0,0) is 

  printSoln 1 $ (fst p1) * (snd p1)
  printSoln 2 $ (fst' p2) * (snd' p2)
